### PR TITLE
(#23) Implement GetByIds in ListingInfoController

### LINKED ISSUE #23: Implement the ListingInformationController GetByIds Method
## As an application developer, I want to retrieve a concept's information listing by specifying its concept IDs, so that I can find its pretty url name.

> **ESTIMATE** TBD

### Acceptance Criteria

```gherkin
Scenario: Get listing information documents with full concept ID list in random order
  Given path '/listing-information', 'get'
    And param  ccode = 'C114967', 'C114964', 'C114972', 'C114963'
  When  method get
  Then  status 200
    And match response == read ('expected result file')

Scenario: listing information documents with only partial concept ID list.
  Given path '/listing-information', 'get'
    And param  ccode = 'C114967', 'C114964'
  When  method get
  Then  status 200
    And match response == read ('expected result file')

Scenario: listing information documents with an invalid concept ID list.
  Given path '/listing-information', 'get'
    And param  ccode = 'chicken'
  When  method get
  Then  status 404

```

### Resources:
*

### Notes

---

### Prerequisites
- Set up API scaffolding - Issue #1 

### Tasks
- 

### Technical Notes
-

### PR DESCRIPTION
**Pull request by sarinapadilla**
* Add GetByIds endpoint to ListingInformationController
* Add GetByIds method to ListingInfoQueryService interface
* Add GetByIds method to ListingInfoQueryService ES implementation
* Update ListingInfo model to remove fields based on requirements
* Update ListingInformationComparer for tests
* Add ResultsMetadataComparer for tests
* Add ListingResultsComparer for tests
* Add ListingInformationController GetByIds tests
* Add ESListingInfoQueryService GetByIds tests
* Add test data for GetByIds tests
* Update folder hierarchy for test data
* Update listing info mapping and data for integration tests
* Update expected data for GetByPrettyUrlName integration tests
* Add integration tests and data for GetByIds
* Closes #23

### CODE REVIEW

File: integration-tests/features/listing-information/getByIds/id-not-found-multiple-recurrent-brain-neoplasm.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (blairlearn) [inline integration-tests/features/listing-information/getByIds/id-not-found-multiple-recurrent-brain-neoplasm.json]: I'm guilty of this myself, but it would be good to have the data file name reflect what's being tested rather than what's in the data (or in this case, what's also not in the data). So for this one, based on the message you're expecting, a better file name might be "id-not-found-multiple-codes.json".  Then you have additional tests for "single-missing-code" and "invalid-code" (chicken).

Arguably `id-not-found.feature` should really just be `errors.feature` since status 409 is "Conflict" rather than "Not Found".

File: src/NCI.OCPL.Api.CTSListingPages/Controllers/ListingInformationController.cs [hunk 2]
```diff
@@ -66,5 +66,45 @@ public async Task<ListingInfo> GetByPrettyUrlName(string prettyUrlName)
...
+        }
+
+        /// <summary>
+        /// Determines whether a given array is null or empty.
+        /// </summary>
+        /// <returns>A boolean.</returns>
+        private static bool IsNullOrEmpty(string[] myStringArray)
+        {
+            return myStringArray == null || myStringArray.Length < 1;
+        }
```
Comment (blairlearn) [inline src/NCI.OCPL.Api.CTSListingPages/Controllers/ListingInformationController.cs]: It would require less code to just put `if( ccode == null || ccode.Length == 0)` in the caller. 😉   Also, this method doesn't actually check whether the individual array elements are valid.

This method should be removed and replaced with an inline check along these lines, but you should also consider detecting/handling the case where _some_ of the IDs are blanks.  It's valid to return an error if _some_ are blank. (It's also valid to only the process the valid ones, but that involves guessing the user's intent.)
```
if( ccode == null || ccode.Length == 0 || Array.TrueForAll(ccode, (code) => string.IsNullOrWhiteSpace(code)))
```

Take a look at [IEnumerable<T>.Any](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.any?view=netcore-2.1#System_Linq_Enumerable_Any__1_System_Collections_Generic_IEnumerable___0__System_Func___0_System_Boolean__)

File: test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Controllers/ListingInformationControllerTest.GetByIds.cs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    public partial class ListingInformationControllerTest
+    {
+        /// <summary>
+        /// Verify correcting handling of invalid c-codes.
+        /// </summary>
+        [Theory]
+        [InlineData(new object[] { null })]
+        [InlineData(new object[] { new string[] { } })]       // Can't use String.Empty because it's a property instead of a constant.
+        public async void GetByIds_InvalidCCodes(string[] ccode)
+        {
```
Comment (blairlearn) [inline test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Controllers/ListingInformationControllerTest.GetByIds.cs]: You'll also need to add tests for the variations on an array of nulls and some elements are null.  Also for the array containing empty strings.  (The comment on the second test input appears on tests elsewhere to explain why an explicit empty string `""` is used instead of `String.Empty`.)

File: src/NCI.OCPL.Api.CTSListingPages/Services/ESListingInfoQueryService.cs [hunk 4]
```diff
@@ -79,17 +80,94 @@ public async Task<ListingInfo> GetByPrettyUrlName(string prettyUrlName)
...
+                        TotalResults = Convert.ToInt32(response.Total),
+                        From = 0
+                    },
+                    Results = listingInfoResults.ToArray()
+                };
+
+                // Log an error
+                String msg = $"Multiple records found when searching for c-code(s) '{String.Join(",", ccodes)}'.";
+                _logger.LogWarning(msg);
+            }
```
Comment (blairlearn) [inline src/NCI.OCPL.Api.CTSListingPages/Services/ESListingInfoQueryService.cs]: These  two nearly identical blocks could be combined with an inner `if` to handle the logging.

But really, this method should simply return an array of of `ListingInfo` objects and let the controller decide what to do with 0, 1, or many and creating a wrapper object.  This method's only concern should be retrieving data from ES, not "packaging" it for the consumer.

File: test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Models/ListingResultsComparer.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                return true;
+            }
+            else if (x == null || y == null)
+            {
+                return false;
+            }
+
+            bool isEqual =
+                new ResultsMetadataComparer().Equals(x.Meta, y.Meta)
+                && AreParamArraysEqual<ListingInfo, ListingInformationComparer>(x.Results, y.Results);
```
Comment (blairlearn) [inline test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Models/ListingResultsComparer.cs]: **DO NOT CHANGE**

This is just a note, rather than something to change. [Linq provides a `SequenceEqual` method on `IEnumerable<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.sequenceequal?view=netcore-2.1#definition) which does the same thing as `AreParamArraysEqual`, except that it's not limited to arrays. No points off for using this method, but I find myself moving in the direction of using the one from Linq.

File: test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Services/ESListingInfoQueryServiceTest.GetByIds.cs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            Assert.Equal("application/json", esContentType);
+            Assert.Equal(HttpMethod.POST, esMethod);
+            Assert.Equal(expectedRequest, requestBody, new JTokenEqualityComparer());
+        }
+
+        public static IEnumerable<object[]> GetByIds_Scenarios = new[]
+        {
+            //new object[] { new GetByIds_NoResults() },
+            //new object[] { new GetByIds_SingleResult() },
+            new object[] { new GetByIds_MultipleResults() }
```
Comment (blairlearn) [inline test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Services/ESListingInfoQueryServiceTest.GetByIds.cs]: Were these meant to be left out?

File: test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Services/ESListingInfoQueryServiceTest.GetByIds.cs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        };
+
+        /// <summary>
+        /// Test to verify handling when no results are returned.
+        /// </summary>
```
Comment (blairlearn) [inline test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Services/ESListingInfoQueryServiceTest.GetByIds.cs]: The comment for this test doesn't match with what the test data claims to be.

File: test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Services/ESListingInfoQueryServiceTest.GetByIds.cs [hunk 8]
```diff
@@ -0,0 +1,137 @@
+using System;
```
Comment (blairlearn) [inline test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Services/ESListingInfoQueryServiceTest.GetByIds.cs:1]: Missing a test for the Elasticsearch connection failing. (e.g. returning only a partial response.)

**Review by blairlearn — CHANGES_REQUESTED** [review]

File: integration-tests/features/listing-information/getByIds/id-not-found-multiple-recurrent-brain-neoplasm.json (hunk 1 above)
Comment (sarinapadilla) [reply integration-tests/features/listing-information/getByIds/id-not-found-multiple-recurrent-brain-neoplasm.json]: Updated! Renamed the test data files to include both what's being tested and what's in the data. Also renamed to `errors.feature` and associated test data files.

File: src/NCI.OCPL.Api.CTSListingPages/Controllers/ListingInformationController.cs (hunk 2 above)
Comment (sarinapadilla) [reply src/NCI.OCPL.Api.CTSListingPages/Controllers/ListingInformationController.cs]: Updated! Removed the unnecessary method and did the check within the controller. Used IEnumarable.Any for checking whether any items are null or empty.

File: test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Controllers/ListingInformationControllerTest.GetByIds.cs (hunk 3 above)
Comment (sarinapadilla) [reply test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Controllers/ListingInformationControllerTest.GetByIds.cs]: Updated! Included a variety of tests.

File: test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Services/ESListingInfoQueryServiceTest.GetByIds.cs (hunk 6 above)
Comment (sarinapadilla) [reply test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Services/ESListingInfoQueryServiceTest.GetByIds.cs]: Nope! Updated to include these tests.

File: test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Services/ESListingInfoQueryServiceTest.GetByIds.cs (hunk 7 above)
Comment (sarinapadilla) [reply test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Services/ESListingInfoQueryServiceTest.GetByIds.cs]: Updated the comment here, and in GetByPrettyUrlName tests.

File: test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Services/ESListingInfoQueryServiceTest.GetByIds.cs (hunk 8 above)
Comment (sarinapadilla) [reply test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Services/ESListingInfoQueryServiceTest.GetByIds.cs:1]: Included a test for this!

File: src/NCI.OCPL.Api.CTSListingPages/Services/ESListingInfoQueryService.cs (hunk 4 above)
Comment (sarinapadilla) [reply src/NCI.OCPL.Api.CTSListingPages/Services/ESListingInfoQueryService.cs]: Updated! With the change to the return object, the controller does the logic for error throwing, while the service just sets up the array of `ListingInfo` objects. The service does log a warning for multiple returned, just in case.

File: integration-tests/features/listing-information/getByIds/id-not-found-multiple-recurrent-brain-neoplasm.json (hunk 1 above)
Comment (blairlearn) [reply integration-tests/features/listing-information/getByIds/id-not-found-multiple-recurrent-brain-neoplasm.json]: The file names still reflect the contents (e.g. `errors-C0000-C0001.json`) instead of the functionality being tested (e.g. `errors-multiple-non-existing-codes.json`).

File: src/NCI.OCPL.Api.CTSListingPages/Controllers/ListingInformationController.cs (hunk 2 above)
Comment (blairlearn) [reply src/NCI.OCPL.Api.CTSListingPages/Controllers/ListingInformationController.cs]: 👍

File: test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Controllers/ListingInformationControllerTest.GetByIds.cs (hunk 3 above)
Comment (blairlearn) [reply test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Controllers/ListingInformationControllerTest.GetByIds.cs]: You still have a comment incorrectly explaining that the empty array is there because String.Empty isn't a constant. The empty array is actually there because you're testing that an empty array is handled correctly.

Also consider whether it's worth adding a test for an array consisting solely of the empty string. (I'm good either way.)

File: src/NCI.OCPL.Api.CTSListingPages/Services/ESListingInfoQueryService.cs (hunk 4 above)
Comment (blairlearn) [reply src/NCI.OCPL.Api.CTSListingPages/Services/ESListingInfoQueryService.cs]: A few things here.

First, you're still doing too much work. Instead of looping through the response.Documents collection and individually allocating the `List<T>` items, you can simply pass the collection into the `List<T>` constructor.  e.g. `new List<ListingInfo>(response.Documents)`.

But, I don't believe there's a reason to go through an intermediate collection.  You're returning an array, so as long as you have a valid response, you should be able to simply `return response.Documents.ToArray();` (if the collection is empty, that should return an empty array).  (The extra allocations are an anti-pattern we've picked up because we're all copying code which was originally written by someone who was new to the .Net platform.)

Also, in the interest of full disclosure: My original feedback was partially incorrect.  If we were still working with the possibility of paged output, you would have needed to return the metadata object as the total number of records is only available from the elasticsearch return. But since we're no longer living in that world, having the query return everything it finds is the correct outcome.

File: src/NCI.OCPL.Api.CTSListingPages/Services/ESListingInfoQueryService.cs [hunk 9]
```diff
@@ -1,5 +1,6 @@
 using System;
```
Comment (blairlearn) [inline src/NCI.OCPL.Api.CTSListingPages/Services/ESListingInfoQueryService.cs:1]: Missing a test for the Elasticsearch connection failing. (e.g. returning only a partial response.)

**Review by blairlearn — CHANGES_REQUESTED** [review]

**Review by blairlearn — CHANGES_REQUESTED** [review]
Missing a "connection failed" test for `ListingInfo.GetByPrettyUrlName()`.  No idea why GitHub wouldn't let me include that in the other comments.

File: src/NCI.OCPL.Api.CTSListingPages/Services/ESListingInfoQueryService.cs (hunk 9 above)
Comment (sarinapadilla) [reply src/NCI.OCPL.Api.CTSListingPages/Services/ESListingInfoQueryService.cs:1]: Added!

File: src/NCI.OCPL.Api.CTSListingPages/Services/ESListingInfoQueryService.cs (hunk 4 above)
Comment (sarinapadilla) [reply src/NCI.OCPL.Api.CTSListingPages/Services/ESListingInfoQueryService.cs]: Updated to have a simpler return of the response.Documents.

File: test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Controllers/ListingInformationControllerTest.GetByIds.cs (hunk 3 above)
Comment (sarinapadilla) [reply test/NCI.OCPL.Api.CTSListingPages.Tests/Tests/Controllers/ListingInformationControllerTest.GetByIds.cs]: Added a few new tests and lots of comments describing the reasons for each!

File: integration-tests/features/listing-information/getByIds/id-not-found-multiple-recurrent-brain-neoplasm.json (hunk 1 above)
Comment (sarinapadilla) [reply integration-tests/features/listing-information/getByIds/id-not-found-multiple-recurrent-brain-neoplasm.json]: Updated the file names to reflect the functionality tested.

**Review by blairlearn — APPROVED** [review]
